server/src/test/java/keywhiz/service/resources/automation/v2/SecretResourceTest.java

```diff
@@ -275,6 +279,145 @@
     assertThat(s3).doesNotContain("secret17");
   }
 
+  @Test public void secretVersionListing_notFound() throws Exception {
+    Request put = clientRequest("/automation/v2/secrets/non-existent/versions/0-0").build();
+    Response httpResponse = mutualSslClient.newCall(put).execute();
+    assertThat(httpResponse.code()).isEqualTo(404);
+  }
+
+  @Test public void secretVersionListing_success() throws Exception {
+    int totalVersions = 6;
+    int sleepInterval = 1000; // Delay so secrets have different creation timestamps
+    List<SecretDetailResponseV2> versions;
+    assertThat(listing()).doesNotContain("secret20");
+
+    // get current time to calculate timestamps off for expiry
+    long now = System.currentTimeMillis() / 1000L;
+
+    // Create secrets 1 second apart, so their version order is deterministic
```

Comment:
Minor thing, but I think it's worth expanding this comment to explain why it's necessary to put a 1 second delay to make them be ordered correctly. Otherwise it kind of sounds like it's a concurrency issue/race condition/etc.